File: server/src/engine/decision.js
const { v4: uuidv4 } = require('uuid');

class DecisionEngine {
  constructor() {
    this.recommendations = [];
  }

  process(alerts) {
    const recommendations = [];

    for (const alert of alerts) {
      if (alert.severity === 'CRITICAL') {
        recommendations.push({
          id: uuidv4(),
          type: 'BLOCK',
          action: `Block transaction ${alert.transaction_id}`,
          reason: alert.message,
          risk_score: 1.0,
          status: 'pending',
          created_at: new Date().toISOString()
        });
      } else if (alert.severity === 'HIGH') {
        recommendations.push({
          id: uuidv4(),
          type: 'REVIEW',
          action: `Review transaction ${alert.transaction_id}`,
          reason: alert.message,
          risk_score: 0.7,
          status: 'pending',
          created_at: new Date().toISOString()
        });
      }
    }

    this.recommendations.push(...recommendations);
    return recommendations;
  }

  calculateRiskScore(alerts) {
    if (!alerts.length) return 0;
    const severityScores = { CRITICAL: 1.0, HIGH: 0.7, MEDIUM: 0.4, LOW: 0.1 };
    const total = alerts.reduce((sum, a) => sum + (severityScores[a.severity] || 0.1), 0);
    return Math.min(total / alerts.length, 1.0);
  }
}

module.exports = DecisionEngine;

File: app/engine/decision.py
from typing import List
from models import Recommendation
import uuid


class DecisionEngine:
    def __init__(self):
        self.recommendations = []
    
    def process(self, alerts: List[dict]) -> List[Recommendation]:
        recommendations = []
        
        for alert in alerts:
            if alert.get("severity") == "CRITICAL":
                recommendations.append(Recommendation(
                    id=str(uuid.uuid4()),
                    type="BLOCK",
                    action=f"Block transaction {alert.get('transaction_id')}",
                    reason=alert.get("message"),
                    risk_score=1.0
                ))
            elif alert.get("severity") == "HIGH":
                recommendations.append(Recommendation(
                    id=str(uuid.uuid4()),
                    type="REVIEW",
                    action=f"Review transaction {alert.get('transaction_id')}",
                    reason=alert.get("message"),
                    risk_score=0.7
                ))
        
        self.recommendations.extend(recommendations)
        return recommendations
    
    def calculate_risk_score(self, alerts: List[dict]) -> float:
        if not alerts:
            return 0.0
        
        severity_scores = {"CRITICAL": 1.0, "HIGH": 0.7, "MEDIUM": 0.4, "LOW": 0.1}
        total = sum(severity_scores.get(a.get("severity", "LOW"), 0) for a in alerts)
        return min(total / len(alerts), 1.0)
